Use `Report` effect instead of `error`

just95 opened the pull request:
<!--
  Have you read our Code of Conduct?
  By filing an issue or pull request, you are expected to comply with it, including treating everyone with respect:
  https://github.com/FreeProving/guidelines/blob/main/CODE_OF_CONDUCT.md
-->

### Issue

<!--
  Link to the issue that this pull request addresses.
  If there is not yet a corresponding issue, please open a new issue and then link to that issue in your pull request.
  Give any additional information that is not covered by the linked issue but might be important for the reviewer.
-->
Closes #38.
Closes #17 by completing pattern matching for `S.NegApp`.
Closes #41 in order to complete pattern matching in `HST.Frontend.ToHSE`.

### Description of the Change

<!--
  Give a short summary of the changes you made to handle the issue.
  Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.
-->

### Alternate Designs

<!-- Explain what other alternates were considered and why the proposed version was selected. -->

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process

<!--
  What process did you follow to verify that your change has the desired effects and has not introduced any regressions?
  Describe the actions you performed (including input you checked, tests you created, commands you ran, etc.), and describe the results you observed.
-->

### Additional Notes

<!-- Is there anything else worth mentioning (e.g. changes by your PR that other contributors have to be aware of, ideas for further enhancements, etc.)? -->

 - Support for `newtype` declarations was added in order to complete pattern matching in `HST.Frontend.FromHSE`. `newtype` declarations can be represented by the existing `DataDecl` nodes.

just95 committed:
Remove `error` calls from `HST.Environment.Renaming` #38 #17

The removal of the catch all clause forces pattern matching
in `HST.Environment.Renaming` to be exhaustive. Explicit
handling of `S.NegApp` expressions fixes the example in #17.
There may be more places where `S.NegApp`s need to be handled.

stu204146-kiel commented on src/lib/HST/CoreAlgorithm.hs:
```diff
@@ -360,35 +379,36 @@ translateApp _              = error "translateApp: Unsupported pattern"
-getConst (S.PWildCard _) = return (S.PWildCard S.NoSrcSpan, [], [])
-getConst _               = error "wrong Pattern in getConst"
+decomposeConPat (S.PWildCard _) = return (S.PWildCard S.NoSrcSpan, [], [])
+decomposeConPat _               = error "wrong Pattern in decomposeConPat"
```
There are still errors in here

stu204146-kiel commented on src/lib/HST/CoreAlgorithm.hs:
```diff
@@ -183,37 +200,30 @@ computeAlts x xs eqs er = do
-getMissingConstrs :: [S.Alt a] -> PM a [Constructor a]
-getMissingConstrs []   = error "getMissingConstrs: empty list"
+getMissingConstrs :: Member (Env a) r => [S.Alt a] -> Sem r [ConEntry a]
+getMissingConstrs []   = error "getMissingConstrs: empty case"
```
Here is also an error that should be replaced by report

just95 committed:
Use `Report` effect in `HST.Application` #38

just95 committed:
Complete pattern matching in `HST.Feature.GuardElimination` #38

just95 committed:
Simplify generation of `let` expressions #38

Since I was documenting the definitions in in
`HST.Feature.GuardElimination` anyway, I've just
rewritten the entire code for the generation of
`let` expressions such that it is easier to
comprehend.

just95 committed:
Make `getMatchName` total #38

It is impossible for a `S.FunBind` to have no `S.Match`es.
We could have lifted `getMatchName` into `Sem` and add the
`Report` effect to report an `Internal` error message.
However, `HST.Feature.GuardElimination` uses `head` already
to determine the arity of the function. Thus, I think that
it is safe to use `head` to get the name of the function as
well.

stu204767 commented on src/lib/HST/Application.hs:
```diff
@@ -103,25 +109,38 @@ hasCons (S.InfixMatch _ p1 _ ps _ _) = any isCons (p1 : ps)
+  fromUnguardedRhs :: Member Report r => S.Rhs a -> Sem r (S.Exp a)
+  fromUnguardedRhs (S.UnGuardedRhs _ expr) = return expr
+  fromUnguardedRhs (S.GuardedRhss _ _) =
+    reportFatal $ Message Internal "Expected unguarded right hand side."
```
```suggestion
    reportFatal $ Message Internal "Expected unguarded right-hand side."
```

stu204767 commented on src/lib/HST/Application.hs:
```diff
@@ -103,25 +109,38 @@ hasCons (S.InfixMatch _ p1 _ ps _ _) = any isCons (p1 : ps)
+    expr <- fromUnguardedRhs rhs
+    return (pat : pats, expr)
+
+  -- | Gets the expressions of the given right-hand side of a rule.
```
```suggestion
  -- | Gets the expression of the given unguarded right-hand side of a rule.
```

stu204767 commented on src/lib/HST/Environment/Renaming.hs:
```diff
@@ -119,20 +117,24 @@ instance Rename (S.Alt s) where
---   Applies the substitution to the expression on the right-hand side.
---   There must be no guards.
+--   Applies the substitution to guards and the expression on the right-hand
+--   sides.
```
```suggestion
--   Applies the substitution to all guard and body expressions of the
--   right-hand side(s).
```

stu204767 commented on src/test/HST/ApplicationTests.hs:
```diff
@@ -141,10 +139,10 @@ testProcessModule = context "processModule" $ do
       , "              a3 -> if otherwise then a3"
       , "                                 else a2"
+      , "            a2 = undefined"
       , "         in a1"
```
```suggestion
      , "        in a1"
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -1,255 +1,269 @@
+  => [S.Exp a] -- ^ Variables to match.
+  -> S.Exp a   -- ^ Expression to use if patterns don't match or the guard is
+               --   not satisfied.
+  -> [GExp a]  -- ^ Patters to match and the corrsponding right-hand sides.
```
```suggestion
  -> [GExp a]  -- ^ Patterns to match and the corresponding right-hand sides.
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -1,255 +1,269 @@
+      decls     = zipWith makeVarBinding varNames (caseExprs ++ [err])
+  return $ S.Let S.NoSrcSpan (S.BDecls S.NoSrcSpan decls) startVarExpr
+
+-- | Creates a function declaration for a @let@ binding of a variable with the
+--   given name to the given expression.
```
That comment is confusing to read. This should be better:
```suggestion
-- | Creates a function declaration for a @let@ binding that binds a variable
--   with the given name to the given expression.
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -1,255 +1,269 @@
+-- | Generates nested case expression for each variable and pattern pair.
+--
+--   @'generateNestedCases' e f [(x₁, p₁), …, (xₙ, pₙ)]@ produces an expression
+--   of the following form.
```
```suggestion
--   of the following form:
```

just95 replied:
If we used a colon here, the sentence would never end with a period.

stu204767 replied:
I don't think this is a problem. I've found [this article](https://www.grammarbook.com/punctuation/colons.asp) and looking at the first example of rule 3, it seems to be correct as well.

just95 replied:
The article also states that "there are no hard and fast rules, except be consistent". The rule that I've adopted from @ichistmeinname (and which we should stick to for consistency's sake) is that a colon is only used if the sentence still ends with a period eventually.

stu204767 replied:
Okay, but I don't think this is a sensible rule if it prevents the usage of colons in cases like this. The only option to add a colon here would be to add a period after the example, which would look weird.
The choice between a colon and a period is a stylistic one in the end. I would definitely choose the colon in this case (and I'm [not the only one](https://english.stackexchange.com/questions/96286/period-or-colon)), but I'm fine with a period as well.

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -1,255 +1,269 @@
+-- @case@ Expressions                                                        --
+-------------------------------------------------------------------------------
+
+-- | Generates nested case expression for each variable and pattern pair.
```
```suggestion
-- | Generates a nested case expression for each variable and pattern pair.
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -1,255 +1,269 @@
+--
+--   > case x₁ of { p₁ -> (… case xₙ of { pₙ -> e ; _ -> f }  …) ; _ -> f }
+generateNestedCases
+  :: S.Exp a              -- ^ Expression to use if all pattern match.
```
```suggestion
  :: S.Exp a              -- ^ Expression to use if all patterns match.
```

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -1,255 +1,269 @@
-  [S.alt p (createCase i next vps), S.alt (S.PWildCard S.NoSrcSpan) next]
+generateNestedCases successExpr failExpr = foldr generateNestedCase successExpr
+ where
+  {- generateNestedCase :: (S.Exp a, S.Pat a) -> S.Exp a -> S.Exp a -}
```
Why is the type signature commented out?

just95 replied:
Since `failExpr` is used in `generateNestedCase` the `a` in the type signature of `generateNestedCase` must be the same `a` as in the type signature of `generateNestedCases`. Due to Haskells type variable scoping rules, it is not possible to specify the type of `geneateNestedCase` without language extensions. Thus, the type signature must be commented out as suggested by our [Haskell Styleguide](https://github.com/FreeProving/guidelines/blob/main/CONTRIBUTING.md#add-type-signatures-for-all-function-declarations).

stu204767 commented on src/lib/HST/Feature/GuardElimination.hs:
```diff
@@ -1,255 +1,269 @@
+-- Guard Elimination                                                         --
+-------------------------------------------------------------------------------
+
+-- | Applies guard elimination on @case@ expression in the given expression.
```
```suggestion
-- | Applies guard elimination on @case@ expressions in the given expression.
```

stu204767 committed:
Apply simple suggestions from my own code review

just95 committed:
Merge branch 'issue-31' into issue-38

just95 committed:
Complete pattern in `applyGEDecl` #38

just95 committed:
Use `Report` effect in `FromHSE` #38

just95 committed:
Complete pattern matching in `ToHSE` #38 #41

Fixes #41 by adding a `OtherDecl` constructor and `OriginalDecl`
type family in favor of the `TypeSig` constructor. The new
constructors allow us to simplify `HST.Frontend.ToHSE` such
that neither `error` nor `reportFatal` error calls are necessary.

just95 committed:
Fix `Main` #38 #41

The changes in 51b73fb1f328835ac71cbe232f5a60ecff50e60a were not
applied to the `Main` module.

just95 committed:
Use `Report` effect in `Optimization` #38

just95 committed:
Add `PatternStack` effect #38

I've noticed that the `matchedPat` field of the
`Environment` is needed in the `Optimization`
module only. Thus, I was able to move the field
to its own state effect and run the handler in
`optimize` in order to provide the effect only
locally.

just95 committed:
Complete pattern matching in `Optimization` #38

just95 committed:
Complete pattern matching in `Renaming` #38

just95 committed:
Complete pattern matching in `CaseCompletion` #38

just95 committed:
Add additional interpretations for `Report` #38

It is sometimes necessary to discard reported messages
or to report pattern matching failures in `do` blocks.
Handlers have been added for those two scenarios.

just95 committed:
Add selectors for pattern names #38

In `HST.CoreAlgorithm` there are functions for getting the
name of a constructor or variable matched by a pattern.
The functions are used in other files as well and have
incomplete pattern matching. Alternative selectors with
complete pattern matching and proper error reporting have
been added to replace `getQNamePat` and `getPVarName`.

just95 committed:
Add error reporting lookup functions #38

When looking up entries in the `Environment` the result is
wrapped with `Maybe` because the type or constructor may not
be in scope. Utility functions have been added that report
a fatal message when `Nothing` is returned.

just95 committed:
Add utility module for commpon predicates #32

The `HST.CoreAlgorithm` module exports predicates for testing
the type of patterns. Since the same functions are needed in
other modules as well, they have been moved to their own module.
Furthermore, pattern matching in the predicates has been completed.

just95 committed:
Add pretty printing functions for names #38

In error messages names often need to be pretty printed.
Utility functions have been added to pretty print names
of the intermediate representation.

just95 committed:
Add pattern to expression conversion function #38

In `HST.CoreAlgorithm` there are partial functions to convert
some patterns to expressions. In order to complete pattern
matching a new function that converts arbitrary patterns to
expressions has been added.

just95 committed:
Complete pattern matching in `CoreAlgorithm` #38

just95 committed:
Format code #38

just95 committed:
Apply suggestions for documentation from review #39

stu204146-kiel commented on src/lib/HST/Effect/PatternStack.hs:
```diff
@@ -0,0 +1,74 @@
+-- | An effect capable of pushing and poping patterns variables have been
+--   matched against to a stack.
+data PatternStack a m b where
+  PushPattern ::S.QName a -> S.Pat a -> PatternStack a m ()
```
It should be a whitespace between the two colons and the type expression
```suggestion
  PushPattern :: S.QName a -> S.Pat a -> PatternStack a m ()
```

just95 replied:
I agree, but Brittany would change it back. We will have to wait for <https://github.com/FreeProving/guidelines/issues/5>.

stu204146-kiel commented on src/lib/HST/Effect/PatternStack.hs:
```diff
@@ -0,0 +1,74 @@
+    :: Sem (PatternStack a ': r) b -> Sem (State (StackMap a) ': r) b
+  patternStackToState = reinterpret \case
+    PushPattern name pat -> modify $ Map.alter (maybeCons pat) name
+    PeekPattern name     -> gets $ fmap head . Map.lookup name
```
Here, the computation can fail if the list is empty. I would catch the error by returning a `Maybe`.

just95 replied:
The computation cannot fail since entries are added and removed with `PushPattern` and `PopPattern` only. `PopPattern` removes the entry from the map if the list is empty via `maybeTail`. Thus, `Map.lookup` (and consequently `PeekPattern`) returns `Nothing` if there is no pattern for `name`. If `Map.lookup` returns a `Just`, the list of patterns is guaranteed to be non-empty.

stu204146-kiel commented on src/lib/HST/Environment/Renaming.hs:
```diff
@@ -37,24 +37,23 @@ class TermSubst c where
+  substitute s (S.NegApp l e    ) = S.NegApp l (substitute s e)
+  substitute s (S.App    l e1 e2) = S.App l (substitute s e1) (substitute s e2)
+  substitute s (S.Lambda l ps e ) = S.Lambda l ps (substitute s e)
+  substitute s (S.Let    l b  e ) = S.Let l b $ substitute s e
```
It is not ensured that name clashes are resolved: Example `let x = 1 in e + x`. If we substitute `e` by `x` we get a name clash

just95 replied:
That's true and a general problem of the `Renaming` module. The issue has been reported as #16 already.

just95 committed:
Add `Data.Map.Strict` to `.hlint.yaml` #38

stu204767 commented on src/lib/HST/Effect/PatternStack.hs:
```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE TemplateHaskell, ScopedTypeVariables #-}
+{-# LANGUAGE LambdaCase, BlockArguments #-}
+
+-- | This module defines an effect for computations that can save patterns
+--   variables have been matched against on a stack.
```
I'm not sure if this is what you meant:
```suggestion
-- | This module defines an effect for computations that can save pattern
--   variables that have been matched against on a stack.
```

just95 replied:
I agree that it is difficult to parse that sentence. What I mean is that the computation can store patterns that variables have been matched against. For example in

```haskell
case xs of
  (x : xs') -> …
```

the pattern `(x : xs')` would be pushed on the stack for the variable `xs`.

stu204767 replied:
Ha, this is an advantage of compound words: You know that separate nouns are separate subjects/objects in a sentence, while you can only guess from the context without them. In this case, I would suggest:
```suggestion
-- | This module defines an effect for computations which can save patterns
--   that variables have been matched against on a stack.
```
I have also updated my other suggestions regarding this sentence structure.

stu204767 commented on src/lib/HST/Effect/PatternStack.hs:
```diff
@@ -0,0 +1,74 @@
+-------------------------------------------------------------------------------
+
+-- | An effect capable of pushing and poping patterns variables have been
+--   matched against to a stack.
```
See the comment above.
```suggestion
-- | An effect capable of pushing and popping patterns that variables have been
--   matched against to a stack.
```

stu204767 commented on src/lib/HST/Environment/LookupOrReport.hs:
```diff
@@ -0,0 +1,61 @@
+-- | This module contains variations of the "HST.Environment" lookup functions
+--   that don't return @Maybe@ values but report fatal error if a data or type
+--   constructor cannot be found in the environment.
```
```suggestion
--   that don't return @Maybe@ values but report a fatal error if a data or
--   type constructor cannot be found in the environment.
```

just95 committed:
Merge branch 'main' into issue-38

stu204767 commented on src/lib/HST/Feature/Optimization.hs:
```diff
@@ -6,63 +6,84 @@ module HST.Feature.Optimization
+optimize = runPatternStack . optimize'
+
+-- | Like 'optimize' but can access a stack of patterns for each local
+--   variable to remember which patterns variables have been matched
```
```suggestion
--   variable to remember the patterns that variables have been matched
```

just95 committed:
Format code #39

stu204767 commented on src/lib/HST/Frontend/FromHSE.hs:
```diff
@@ -25,201 +44,437 @@ data HSE
+  skipNotSupported "Non-variable pattern bindings"
+  return $ S.OtherDecl (transformSrcSpan s) decl
+
+-- Type classes and type class instances ar enot supported. The user is
```
```suggestion
-- Type classes and type class instances are not supported. The user is
```

stu204767 commented on src/lib/HST/Frontend/FromHSE.hs:
```diff
@@ -25,201 +44,437 @@ data HSE
+  skipNotSupported "Type class instances"
+  return $ S.OtherDecl (transformSrcSpan s) decl
+
+-- GADTs and pattern synonyms are not supported. The user us explicitly
```
```suggestion
-- GADTs and pattern synonyms are not supported. The user is explicitly
```

stu204767 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -36,53 +42,90 @@ type family TypeExp a
---   unsupported declarations) from the original module.
-data Module a = Module (Maybe (ModuleName a)) [Decl a]
+--   The pattern matching compiler only needs to know the declarations of
+--   the module. All other information is stored in the 'OriginalModuleHead'.
```
```suggestion
--   The pattern matching compiler only needs to know the declarations and the
--   name of the module. All other information is stored in the
--   'OriginalModuleHead'.
```

stu204767 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -36,53 +42,90 @@ type family TypeExp a
 --   The only supported kind of pattern bindings, variable patterns, are
 --   represented by function bindings.
-data Decl a = DataDecl (Name a) [ConDecl a]
-            | TypeSig (SrcSpan a) [Name a] (TypeExp a)
-            | FunBind (SrcSpan a) [Match a]
+data Decl a
+  = DataDecl (SrcSpan a) (OriginalDecl a) (Name a) [ConDecl a]
+    -- ^ A @data@ or @newtype@ declaration.
+    --
+    --   The 'OriginalDecl' is the original declaration the data type
+    --   declaration has been generated from. This field is needed to
+    --   convert the AST back to the original AST since data type and
+    --   constructor declarations don't contain sufficient information
+    --   to reconstruct the original declaration.
+  | FunBind (SrcSpan a) [Match a]
+    -- ^ A function declaration.
+    --
+    --   We do not need to store the 'OriginalDecl' of the function
+    --   declaration since the AST provides all information that is
+    --   required to reconstruct the original declaration.
+  | OtherDecl (SrcSpan a) (OriginalDecl a)
+    -- ^ An unsupported declaration.
+    --
+    --   This constructor is needed to skip declarations that cannot
+    --   be transformed by the pattern matching compiler.
 deriving instance EqAST a => Eq (Decl a)
 deriving instance ShowAST a => Show (Decl a)
 
+-- | Gets the source span information of a declaration.
+instance HasSrcSpan Decl where
+  getSrcSpan (DataDecl srcSpan _ _ _) = srcSpan
+  getSrcSpan (FunBind   srcSpan _   ) = srcSpan
+  getSrcSpan (OtherDecl srcSpan _   ) = srcSpan
+
 -------------------------------------------------------------------------------
 -- Data Type Declarations                                                    --
 -------------------------------------------------------------------------------
 
 -- | A data constructor. Does not include a source span and should not be
 --   transformed back.
```
The comment needs to be adjusted as well:
```suggestion
-- | A data constructor. Should not be transformed back.
```

just95 replied:
```suggestion
-- | A data constructor.
--
--   Data constructors should not be converted back. The original
--   constructor declaration should be part of the 'OriginalDecl'
--   of a 'DataDecl'.
```

stu204767 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -170,34 +258,116 @@ data Pat a = PVar (SrcSpan a) (Name a)
+  getSrcSpan (PList  srcSpan _       ) = srcSpan
+  getSrcSpan (PWildCard srcSpan      ) = srcSpan
+
+-- | Converts a pattern to an expressions.
```
```suggestion
-- | Converts a pattern to an expression.
```

stu204767 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -170,34 +258,116 @@ data Pat a = PVar (SrcSpan a) (Name a)
   deriving (Eq, Ord)
 deriving instance ShowAST a => Show (QName a)
 
+-- | Wraps a name with 'UnQual'.
+--
+--   The additional source span information is taken from the given name.
+unQual :: Name a -> QName a
+unQual name = UnQual (getSrcSpan name) name
+
+-- | Wraps a name with 'Special'.
+--
+--   The additional source span information is taken from the given name.
+special :: SpecialCon a -> QName a
+special specialCon = Special (getSrcSpan specialCon) specialCon
```
Why are these individual function when they could be part of the `QNameLike` instances for `Name`s and `SpecialCon`s?

just95 replied:
I just thought that

```haskell
patToExp (PVar srcSpan name) = Var srcSpan (unQual name)
```

would be more readable than

```haskell
patToExp (PVar srcSpan name) = Var srcSpan (toQName name)
```

stu204767 replied:
Hm, there is no difference in readability to me. What do you think, @stu204146-kiel?

stu204767 commented on src/lib/HST/Frontend/ToHSE.hs:
```diff
@@ -13,38 +13,46 @@ module HST.Frontend.ToHSE where
+--   module into the @haskell-src-exts@ representation of a Haskell module.
+--
+--   The module head is restored from the original module head. The module
+--   name field does not affect the name of the resulting module..
```
```suggestion
--   name field does not affect the name of the resulting module.
```

stu204767 commented on src/lib/HST/Util/PrettyName.hs:
```diff
@@ -0,0 +1,57 @@
+-- | Helper functions for 'PrettyName' instance of 'S.Name' and 'S.QName'
+--   that pretty prints a name but adds a prefix.
+--
+--    The prefix is added before the identifier or symbol. In case of symbolic
+--    names, the prefix is within the parentheses.
```
```suggestion
--   The prefix is added before the identifier or symbol. In case of symbolic
--   names, the prefix is within the parentheses.
```

stu204767 committed:
Apply simple suggestions from my own code review

just95 committed:
Apply suggestions from code review #39

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

just95 committed:
Apply suggestions from code review #39

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

stu204767 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -114,8 +115,11 @@ instance HasSrcSpan Decl where
---   transformed back.
+-- | A data constructor.
+--
+--   Data constructors should not be converted back. The original
+--   constructor declaration should be part of the 'OriginalDecl'
+--   of a 'DataDecl'.
```
Did you use another limit than 80 characters inadvertently or deliberately?
```suggestion
--   Data constructors should not be converted back. The original constructor
--   declaration should be part of the 'OriginalDecl' of a 'DataDecl'.
```

just95 committed:
Update src/lib/HST/Frontend/Syntax.hs #39

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

stu204767 commented on src/lib/HST/Frontend/FromHSE.hs:
```diff
@@ -25,201 +44,437 @@ data HSE
+-- Error messages                                                           --
+------------------------------------------------------------------------------
+
+-- | Reports a fatal error that the given feature is not supported.
+notSupported
+  :: Member Report r
+  => String -- ^ The name of the feature (plural) that is not supported.
+  -> Sem r b
+notSupported feature =
+  reportFatal $ Message Error $ feature ++ " are not supported!"
+
+-- | Informs the user that the given feature is not supported and the
+--   corresponding AST node will be skipped.
+skipNotSupported
+  :: Member Report r
+  => String -- ^ The name of the feature (plural) that is not supported.
+  -> Sem r ()
+skipNotSupported feature =
+  report $ Message Info $ feature ++ " are not supported and will be skipped!"
```
These functions will be identical for the GHC transformation. I will copy them for now into `FromGHC.hs`, so I can start working with using `report` instead of `error` in the GHC modules, but they should be moved to another module, maybe `HST.Effect.Report`.

just95 replied:
See also #40.

stu204146-kiel commented on src/lib/HST/Frontend/FromHSE.hs:
```diff
@@ -25,201 +44,437 @@ data HSE
-transformBinds _ = error "Implicit bindings are not supported"
+transformBinds
+  :: Member Report r => HSE.Binds HSE.SrcSpanInfo -> Sem r (S.Binds HSE)
+transformBinds (HSE.BDecls s decls) = do
```
Here is a superfluous ``do``

just95 committed:
Remove superfluous `do` #39

just95 committed:
Merge branch 'main' into issue-38
